In [8]:
import pandas as pd
import numpy as np 
from sklearn.preprocessing import MinMaxScaler

In [10]:
df = pd.read_csv("/Users/siddhantgond/Desktop/6THSEM/Project_Elective/dataset_extracted_80_20.csv")

/var/folders/p0/fzpcz2t93td2j063ysbwphbr0000gn/T/ipykernel_2083/910623156.py:1: DtypeWarning: Columns (6,11,13,14,15,16,17,31,32,34,45,51,54,55) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("/Users/siddhantgond/Desktop/6THSEM/Project_Elective/dataset_extracted_80_20.csv")


In [14]:
df.head()

frame.time    ip.src_host    ip.dst_host arp.dst.proto_ipv4  \
0   2021 11:44:10.081753000   192.168.0.128  192.168.0.101                  0   
1   2021 11:44:10.162218000   192.168.0.101  192.168.0.128                  0   
2   2021 11:44:10.162271000   192.168.0.128  192.168.0.101                  0   
3   2021 11:44:10.162641000   192.168.0.128  192.168.0.101                  0   
4   2021 11:44:10.166132000   192.168.0.101  192.168.0.128                  0   

   arp.opcode  arp.hw.size arp.src.proto_ipv4  icmp.checksum  icmp.seq_le  \
0         0.0          0.0                  0            0.0          0.0   
1         0.0          0.0                  0            0.0          0.0   
2         0.0          0.0                  0            0.0          0.0   
3         0.0          0.0                  0            0.0          0.0   
4         0.0          0.0                  0            0.0          0.0   

   icmp.transmit_timestamp  ...  mqtt.proto_len mqtt.protoname  \
0                      0.0  ...             0.0              0   
1                      0.0  ...             4.0           MQTT   
2                      0.0  ...             0.0              0   
3                      0.0  ...             0.0              0   
4                      0.0  ...             0.0              0   

                 mqtt.topic mqtt.topic_len mqtt.ver mbtcp.len mbtcp.trans_id  \
0                         0            0.0      0.0       0.0            0.0   
1                         0            0.0      4.0       0.0            0.0   
2                         0            0.0      0.0       0.0            0.0   
3                         0            0.0      0.0       0.0            0.0   
4  Temperature_and_Humidity           24.0      0.0       0.0            0.0   

  mbtcp.unit_id  Attack_label  Attack_type  
0           0.0             0       Normal  
1           0.0             0       Normal  
2           0.0             0       Normal  
3           0.0             0       Normal  
4           0.0             0       Normal  

[5 rows x 63 columns]

In [16]:
df.shape

(71401, 63)

In [20]:
df["Attack_type"].unique()

array(['Normal', 'MITM', 'Uploading', 'Ransomware', 'SQL_injection',
       'DDoS_HTTP', 'DDoS_TCP', 'Password', 'Port_Scanning',
       'Vulnerability_scanner', 'Backdoor', 'XSS', 'Fingerprinting',
       'DDoS_UDP', 'DDoS_ICMP'], dtype=object)

In [34]:
df["Attack_type"].value_counts()

Attack_type
Normal                   10000
Uploading                 5000
Ransomware                5000
SQL_injection             5000
DDoS_HTTP                 5000
DDoS_TCP                  5000
Password                  5000
Port_Scanning             5000
Vulnerability_scanner     5000
Backdoor                  5000
XSS                       5000
DDoS_UDP                  5000
DDoS_ICMP                 5000
Fingerprinting            1001
MITM                       400
Name: count, dtype: int64

In [36]:
df.columns

Index(['frame.time', 'ip.src_host', 'ip.dst_host', 'arp.dst.proto_ipv4',
       'arp.opcode', 'arp.hw.size', 'arp.src.proto_ipv4', 'icmp.checksum',
       'icmp.seq_le', 'icmp.transmit_timestamp', 'icmp.unused',
       'http.file_data', 'http.content_length', 'http.request.uri.query',
       'http.request.method', 'http.referer', 'http.request.full_uri',
       'http.request.version', 'http.response', 'http.tls_port', 'tcp.ack',
       'tcp.ack_raw', 'tcp.checksum', 'tcp.connection.fin',
       'tcp.connection.rst', 'tcp.connection.syn', 'tcp.connection.synack',
       'tcp.dstport', 'tcp.flags', 'tcp.flags.ack', 'tcp.len', 'tcp.options',
       'tcp.payload', 'tcp.seq', 'tcp.srcport', 'udp.port', 'udp.stream',
       'udp.time_delta', 'dns.qry.name', 'dns.qry.name.len', 'dns.qry.qu',
       'dns.qry.type', 'dns.retransmission', 'dns.retransmit_request',
       'dns.retransmit_request_in', 'mqtt.conack.flags',
       'mqtt.conflag.cleansess', 'mqtt.conflags', 'mqtt.hdrflags', 'mqtt.len

In [46]:
import psutil
import GPUtil
import platform
import os
import subprocess

def get_system_info():
    system_info = {
        "System": platform.system(),
        "Node Name": platform.node(),
        "Release": platform.release(),
        "Version": platform.version(),
        "Machine": platform.machine(),
        "Processor": platform.processor(),
        "Architecture": platform.architecture()[0],
    }
    return system_info

def get_ram_info():
    ram = psutil.virtual_memory()
    swap = psutil.swap_memory()
    ram_info = {
        "Total RAM (GB)": round(ram.total / (1024 ** 3), 2),
        "Available RAM (GB)": round(ram.available / (1024 ** 3), 2),
        "Used RAM (GB)": round(ram.used / (1024 ** 3), 2),
        "RAM Usage (%)": ram.percent,
        "Total Swap (GB)": round(swap.total / (1024 ** 3), 2),
        "Used Swap (GB)": round(swap.used / (1024 ** 3), 2),
        "Swap Usage (%)": swap.percent,
    }
    return ram_info

def get_gpu_info():
    gpus = GPUtil.getGPUs()
    gpu_info = []
    for gpu in gpus:
        gpu_info.append({
            "GPU Name": gpu.name,
            "GPU Driver": gpu.driver,
            "Total Memory (GB)": round(gpu.memoryTotal / 1024, 2),
            "Used Memory (GB)": round(gpu.memoryUsed / 1024, 2),
            "Free Memory (GB)": round(gpu.memoryFree / 1024, 2),
            "GPU Load (%)": gpu.load * 100,
            "Temperature (°C)": gpu.temperature,
        })
    return gpu_info

def get_storage_info():
    partitions = psutil.disk_partitions()
    storage_info = {}
    for partition in partitions:
        try:
            usage = psutil.disk_usage(partition.mountpoint)
            storage_info[partition.device] = {
                "Mount Point": partition.mountpoint,
                "File System": partition.fstype,
                "Total Space (GB)": round(usage.total / (1024 ** 3), 2),
                "Used Space (GB)": round(usage.used / (1024 ** 3), 2),
                "Free Space (GB)": round(usage.free / (1024 ** 3), 2),
                "Usage (%)": usage.percent,
            }
        except PermissionError:
            continue
    return storage_info

def main():
    output = []
    output.append("System Information:")
    for key, value in get_system_info().items():
        output.append(f"{key}: {value}")
    
    output.append("\nRAM Information:")
    for key, value in get_ram_info().items():
        output.append(f"{key}: {value}")
    
    output.append("\nGPU Information:")
    for gpu in get_gpu_info():
        for key, value in gpu.items():
            output.append(f"{key}: {value}")
        output.append("")
    
    output.append("Storage Information:")
    for device, info in get_storage_info().items():
        output.append(f"Device: {device}")
        for key, value in info.items():
            output.append(f"  {key}: {value}")
        output.append("")
    
    with open("system_info.txt", "w") as file:
        file.write("\n".join(output))

if __name__ == "__main__":
    main()

In [84]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71401 entries, 0 to 71400
Data columns (total 63 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   frame.time                 71401 non-null  object 
 1   ip.src_host                71401 non-null  object 
 2   ip.dst_host                71401 non-null  object 
 3   arp.dst.proto_ipv4         71401 non-null  object 
 4   arp.opcode                 71401 non-null  float64
 5   arp.hw.size                71401 non-null  float64
 6   arp.src.proto_ipv4         71401 non-null  object 
 7   icmp.checksum              71401 non-null  float64
 8   icmp.seq_le                71401 non-null  float64
 9   icmp.transmit_timestamp    71401 non-null  float64
 10  icmp.unused                71401 non-null  float64
 11  http.file_data             71401 non-null  object 
 12  http.content_length        71401 non-null  float64
 13  http.request.uri.query     71401 non-null  obj

In [86]:
# Droping unnecessary object columns (timestamps, full URIs, messages, etc.) to reduce the memory usuage
drop_cols = ["frame.time", "http.file_data", "http.request.full_uri", "http.referer", "mqtt.msg", "tcp.payload"]
df = df.drop(columns=drop_cols, errors='ignore')

In [88]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71401 entries, 0 to 71400
Data columns (total 57 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   ip.src_host                71401 non-null  object 
 1   ip.dst_host                71401 non-null  object 
 2   arp.dst.proto_ipv4         71401 non-null  object 
 3   arp.opcode                 71401 non-null  float64
 4   arp.hw.size                71401 non-null  float64
 5   arp.src.proto_ipv4         71401 non-null  object 
 6   icmp.checksum              71401 non-null  float64
 7   icmp.seq_le                71401 non-null  float64
 8   icmp.transmit_timestamp    71401 non-null  float64
 9   icmp.unused                71401 non-null  float64
 10  http.content_length        71401 non-null  float64
 11  http.request.uri.query     71401 non-null  object 
 12  http.request.method        71401 non-null  object 
 13  http.request.version       71401 non-null  obj

In [90]:
#For Attack Type doing Label Encoding
le = LabelEncoder()
df["Attack_type"] = le.fit_transform(df["Attack_type"])

NameError: name 'LabelEncoder' is not defined

In [92]:
from sklearn.preprocessing import LabelEncoder, OneHotEncoder

In [94]:
#For Attack Type doing Label Encoding
le = LabelEncoder()
df["Attack_type"] = le.fit_transform(df["Attack_type"])

In [96]:
#Applying ONE for protocols and request methods
one_hot_cols = ["ip.src_host", "ip.dst_host", "arp.dst.proto_ipv4", "arp.src.proto_ipv4", 
                "http.request.method", "http.request.version", "mqtt.protoname", "mqtt.topic"]

df = pd.get_dummies(df, columns=one_hot_cols, drop_first=True)

# Save Encoded Dataset
df.to_csv("encoded_dataset.csv", index=False)
print("Encoding Complete. Data Saved.")

Encoding Complete. Data Saved.


In [100]:
df_encoded = pd.read_csv("encoded_dataset.csv")

/var/folders/p0/fzpcz2t93td2j063ysbwphbr0000gn/T/ipykernel_2083/951095416.py:1: DtypeWarning: Columns (7,21,23,34) have mixed types. Specify dtype option on import or set low_memory=False.
  df_encoded = pd.read_csv("encoded_dataset.csv")


In [101]:
df_encoded

arp.opcode  arp.hw.size  icmp.checksum  icmp.seq_le  \
0             0.0          0.0            0.0          0.0   
1             0.0          0.0            0.0          0.0   
2             0.0          0.0            0.0          0.0   
3             0.0          0.0            0.0          0.0   
4             0.0          0.0            0.0          0.0   
...           ...          ...            ...          ...   
71396         0.0          0.0        12057.0      57217.0   
71397         0.0          0.0        11289.0      57220.0   
71398         0.0          0.0         7705.0      57234.0   
71399         0.0          0.0         3865.0      57249.0   
71400         0.0          0.0          281.0      57263.0   

       icmp.transmit_timestamp  icmp.unused  http.content_length  \
0                          0.0          0.0                  0.0   
1                          0.0          0.0                  0.0   
2                          0.0          0.0                  0.0   
3                          0.0          0.0                  0.0   
4                          0.0          0.0                  0.0   
...                        ...          ...                  ...   
71396                      0.0          0.0                  0.0   
71397                      0.0          0.0                  0.0   
71398                      0.0          0.0                  0.0   
71399                      0.0          0.0                  0.0   
71400                      0.0          0.0                  0.0   

      http.request.uri.query  http.response  http.tls_port  ...  \
0                        0.0            0.0            0.0  ...   
1                        0.0            0.0            0.0  ...   
2                        0.0            0.0            0.0  ...   
3                        0.0            0.0            0.0  ...   
4                        0.0            0.0            0.0  ...   
...                      ...            ...            ...  ...   
71396                    0.0            0.0            0.0  ...   
71397                    0.0            0.0            0.0  ...   
71398                    0.0            0.0            0.0  ...   
71399                    0.0            0.0            0.0  ...   
71400                    0.0            0.0            0.0  ...   

       http.request.version_0.0  http.request.version_HTTP/1.0  \
0                          True                          False   
1                          True                          False   
2                          True                          False   
3                          True                          False   
4                          True                          False   
...                         ...                            ...   
71396                     False                          False   
71397                     False                          False   
71398                     False                          False   
71399                     False                          False   
71400                     False                          False   

       http.request.version_HTTP/1.1  \
0                              False   
1                              False   
2                              False   
3                              False   
4                              False   
...                              ...   
71396                          False   
71397                          False   
71398                          False   
71399                          False   
71400                          False   

       http.request.version_name=a><input name=i value=XSS>&lt;script>alert('Vulnerable')</script> HTTP/1.1  \
0                                                  False                                                      
1                                                  False                                                      
2                                

## **Conclusion**
During the feature encoding process, it was observed that applying **One-Hot Encoding** significantly increased the dataset size to **4.2GB**, making it impractical for IoT-based Intrusion Detection Systems (IDS). IoT datasets typically require lightweight models and efficient storage to enable real-time analysis on resource-constrained devices.  

### **Key Takeaways:**
- **Label Encoding** for categorical target variables (e.g., `Attack_type`) remains feasible.  
- **One-Hot Encoding** is **not suitable** due to excessive memory consumption.  
- Feature selection should focus on **numeric features** with high relevance to intrusion detection.  
- Instead of expanding categorical features, **embedding-based approaches** or **hashing techniques** may be explored if categorical data needs to be retained.  

To maintain a scalable and efficient IDS, we will prioritise **numeric feature selection** and **dimensionality reduction techniques** instead of full categorical encoding.  


In [163]:
df = pd.read_csv("/Users/siddhantgond/Desktop/6THSEM/Project_Elective/dataset_extracted_80_20.csv")

/var/folders/p0/fzpcz2t93td2j063ysbwphbr0000gn/T/ipykernel_2083/910623156.py:1: DtypeWarning: Columns (6,11,13,14,15,16,17,31,32,34,45,51,54,55) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("/Users/siddhantgond/Desktop/6THSEM/Project_Elective/dataset_extracted_80_20.csv")


In [110]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71401 entries, 0 to 71400
Data columns (total 63 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   frame.time                 71401 non-null  object 
 1   ip.src_host                71401 non-null  object 
 2   ip.dst_host                71401 non-null  object 
 3   arp.dst.proto_ipv4         71401 non-null  object 
 4   arp.opcode                 71401 non-null  float64
 5   arp.hw.size                71401 non-null  float64
 6   arp.src.proto_ipv4         71401 non-null  object 
 7   icmp.checksum              71401 non-null  float64
 8   icmp.seq_le                71401 non-null  float64
 9   icmp.transmit_timestamp    71401 non-null  float64
 10  icmp.unused                71401 non-null  float64
 11  http.file_data             71401 non-null  object 
 12  http.content_length        71401 non-null  float64
 13  http.request.uri.query     71401 non-null  obj

In [112]:
drop_cols = ["frame.time", "http.file_data", "http.request.full_uri", "http.referer", "mqtt.msg", "tcp.payload"]
df = df.drop(columns=drop_cols, errors='ignore')

In [116]:
# Display remaining categorical columns
object_cols = df.select_dtypes(include=['object']).columns

object_cols

Index(['ip.src_host', 'ip.dst_host', 'arp.dst.proto_ipv4',
       'arp.src.proto_ipv4', 'http.request.uri.query', 'http.request.method',
       'http.request.version', 'tcp.options', 'tcp.srcport',
       'mqtt.conack.flags', 'mqtt.protoname', 'mqtt.topic', 'Attack_type'],
      dtype='object')

In [118]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71401 entries, 0 to 71400
Data columns (total 57 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   ip.src_host                71401 non-null  object 
 1   ip.dst_host                71401 non-null  object 
 2   arp.dst.proto_ipv4         71401 non-null  object 
 3   arp.opcode                 71401 non-null  float64
 4   arp.hw.size                71401 non-null  float64
 5   arp.src.proto_ipv4         71401 non-null  object 
 6   icmp.checksum              71401 non-null  float64
 7   icmp.seq_le                71401 non-null  float64
 8   icmp.transmit_timestamp    71401 non-null  float64
 9   icmp.unused                71401 non-null  float64
 10  http.content_length        71401 non-null  float64
 11  http.request.uri.query     71401 non-null  object 
 12  http.request.method        71401 non-null  object 
 13  http.request.version       71401 non-null  obj

In [120]:
le = LabelEncoder()
df["Attack_type"] = le.fit_transform(df["Attack_type"])

In [122]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71401 entries, 0 to 71400
Data columns (total 57 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   ip.src_host                71401 non-null  object 
 1   ip.dst_host                71401 non-null  object 
 2   arp.dst.proto_ipv4         71401 non-null  object 
 3   arp.opcode                 71401 non-null  float64
 4   arp.hw.size                71401 non-null  float64
 5   arp.src.proto_ipv4         71401 non-null  object 
 6   icmp.checksum              71401 non-null  float64
 7   icmp.seq_le                71401 non-null  float64
 8   icmp.transmit_timestamp    71401 non-null  float64
 9   icmp.unused                71401 non-null  float64
 10  http.content_length        71401 non-null  float64
 11  http.request.uri.query     71401 non-null  object 
 12  http.request.method        71401 non-null  object 
 13  http.request.version       71401 non-null  obj

In [126]:
df.head(15)

ip.src_host    ip.dst_host arp.dst.proto_ipv4  arp.opcode  arp.hw.size  \
0   192.168.0.128  192.168.0.101                  0         0.0          0.0   
1   192.168.0.101  192.168.0.128                  0         0.0          0.0   
2   192.168.0.128  192.168.0.101                  0         0.0          0.0   
3   192.168.0.128  192.168.0.101                  0         0.0          0.0   
4   192.168.0.101  192.168.0.128                  0         0.0          0.0   
5   192.168.0.128  192.168.0.101                  0         0.0          0.0   
6   192.168.0.101  192.168.0.128                  0         0.0          0.0   
7   192.168.0.128  192.168.0.101                  0         0.0          0.0   
8   192.168.0.101  192.168.0.128                  0         0.0          0.0   
9   192.168.0.128  192.168.0.101                  0         0.0          0.0   
10  192.168.0.101  192.168.0.128                  0         0.0          0.0   
11  192.168.0.128  192.168.0.101                  0         0.0          0.0   
12  192.168.0.128  192.168.0.101                  0         0.0          0.0   
13  192.168.0.101  192.168.0.128                  0         0.0          0.0   
14  192.168.0.128  192.168.0.101                  0         0.0          0.0   

   arp.src.proto_ipv4  icmp.checksum  icmp.seq_le  icmp.transmit_timestamp  \
0                   0            0.0          0.0                      0.0   
1                   0            0.0          0.0                      0.0   
2                   0            0.0          0.0                      0.0   
3                   0            0.0          0.0                      0.0   
4                   0            0.0          0.0                      0.0   
5                   0            0.0          0.0                      0.0   
6                   0            0.0          0.0                      0.0   
7                   0            0.0          0.0                      0.0   
8                   0            0.0          0.0                      0.0   
9                   0            0.0          0.0                      0.0   
10                  0            0.0          0.0                      0.0   
11                  0            0.0          0.0                      0.0   
12                  0            0.0          0.0                      0.0   
13                  0            0.0          0.0                      0.0   
14                  0            0.0          0.0                      0.0   

    icmp.unused  ...  mqtt.proto_len mqtt.protoname                mqtt.topic  \
0           0.0  ...             0.0              0                         0   
1           0.0  ...             4.0           MQTT                         0   
2           0.0  ...             0.0              0                         0   
3           0.0  ...             0.0              0                         0   
4           0.0  ...             0.0              0  Temperature_and_Humidity   
5           0.0  ...             0.0              0                         0   
6           0.0  ...             0.0              0                         0   
7           0.0  ...             0.0              0                         0   
8           0.0  ...             0.0              0                         0   
9           0.0  ...             0.0              0                         0   
10          0.0  ...             4.0           MQTT                         0   
11          0.0  ...             0.0              0                         0   
12          0.0  ...             0.0              0                         0   
13          0.0  ...             0.0              0                         0   
14          0.0  ...             0.0              0                         0   

   mqtt.topic_len  mqtt.ver  mbtcp.len  mbtcp.trans_id  mbtcp.unit_id  \
0             0.0       0.0        0.0             0.0            0.0   
1             0.0       4.0      

In [128]:
df["Attack_type"].unique()

array([ 7,  6, 12, 10, 11,  1,  3,  8,  9, 13,  0, 14,  5,  4,  2])

In [130]:
# Converting float64 to float32 to reduce memory usage
float_cols = df.select_dtypes(include=['float64']).columns
df[float_cols] = df[float_cols].astype('float32')

# Converting int64 to int32 (if values fit)
int_cols = df.select_dtypes(include=['int64']).columns
df[int_cols] = df[int_cols].astype('int32')

# Saving Optimised Dataset
df.to_csv("optimised_dataset.csv", index=False)
print("Optimised Dataset Saved. Reduced Memory Usage.")

Optimised Dataset Saved. Reduced Memory Usage.


In [132]:
df_optimised = pd.read_csv("optimised_dataset.csv")

/var/folders/p0/fzpcz2t93td2j063ysbwphbr0000gn/T/ipykernel_2083/1006910984.py:1: DtypeWarning: Columns (5,11,12,13,27,29,40,48,49) have mixed types. Specify dtype option on import or set low_memory=False.
  df_optimised = pd.read_csv("optimised_dataset.csv")


In [134]:
df_optimised

ip.src_host    ip.dst_host arp.dst.proto_ipv4  arp.opcode  \
0        192.168.0.128  192.168.0.101                  0         0.0   
1        192.168.0.101  192.168.0.128                  0         0.0   
2        192.168.0.128  192.168.0.101                  0         0.0   
3        192.168.0.128  192.168.0.101                  0         0.0   
4        192.168.0.101  192.168.0.128                  0         0.0   
...                ...            ...                ...         ...   
71396   151.73.124.223  192.168.0.128                  0         0.0   
71397    10.114.211.51  192.168.0.128                  0         0.0   
71398   165.159.32.231  192.168.0.128                  0         0.0   
71399  117.108.235.108  192.168.0.128                  0         0.0   
71400    78.114.69.225  192.168.0.128                  0         0.0   

       arp.hw.size arp.src.proto_ipv4  icmp.checksum  icmp.seq_le  \
0              0.0                  0            0.0          0.0   
1              0.0                  0            0.0          0.0   
2              0.0                  0            0.0          0.0   
3              0.0                  0            0.0          0.0   
4              0.0                  0            0.0          0.0   
...            ...                ...            ...          ...   
71396          0.0                0.0        12057.0      57217.0   
71397          0.0                0.0        11289.0      57220.0   
71398          0.0                0.0         7705.0      57234.0   
71399          0.0                0.0         3865.0      57249.0   
71400          0.0                0.0          281.0      57263.0   

       icmp.transmit_timestamp  icmp.unused  ...  mqtt.proto_len  \
0                          0.0          0.0  ...             0.0   
1                          0.0          0.0  ...             4.0   
2                          0.0          0.0  ...             0.0   
3                          0.0          0.0  ...             0.0   
4                          0.0          0.0  ...             0.0   
...                        ...          ...  ...             ...   
71396                      0.0          0.0  ...             0.0   
71397                      0.0          0.0  ...             0.0   
71398                      0.0          0.0  ...             0.0   
71399                      0.0          0.0  ...             0.0   
71400                      0.0          0.0  ...             0.0   

      mqtt.protoname                mqtt.topic mqtt.topic_len  mqtt.ver  \
0                  0                         0            0.0       0.0   
1               MQTT                         0            0.0       4.0   
2                  0                         0            0.0       0.0   
3                  0                         0            0.0       0.0   
4                  0  Temperature_and_Humidity           24.0       0.0   
...              ...                       ...            ...       ...   
71396            0.0                       0.0            0.0       0.0   
71397            0.0                       0.0            0.0       0.0   
71398            0.0                       0.0            0.0       0.0   
71399            0.0                       0.0            0.0       0.0   
71400            0.0                       0.0            0.0       0.0   

       mbtcp.len  mbtcp.trans_id  mbtcp.unit_id  Attack_label  Attack_type  
0            0.0             0.0            0.0             0            7  
1            0.0             0.0            0.0             0            7  
2            0.0             0.0            0.0             0            7  
3            0.0             0.0            0.0             0            7  
4            0.0             0.0            0.0             0            7  
...          ...             ...            ...           ...          ...  
71396        0.0             0.0            0.0             

### Feature selection for Object columns

In [139]:
# Extracting object (categorical) columns
cat_features = df.select_dtypes(include=['object']).columns.tolist()
print("Categorical Features:", cat_features)

# Saving extracted categorical features for reference
df[cat_features].to_csv("categorical_features.csv", index=False)

Categorical Features: ['ip.src_host', 'ip.dst_host', 'arp.dst.proto_ipv4', 'arp.src.proto_ipv4', 'http.request.uri.query', 'http.request.method', 'http.request.version', 'tcp.options', 'tcp.srcport', 'mqtt.conack.flags', 'mqtt.protoname', 'mqtt.topic']


In [141]:
from sklearn.preprocessing import LabelEncoder

df_encoded = df.copy()
le = LabelEncoder()

# Encode categorical features
for col in cat_features:
    df_encoded[col] = le.fit_transform(df[col])

# Save encoded dataset
df_encoded.to_csv("encoded_dataset.csv", index=False)
print("Categorical Features Encoded Successfully.")

TypeError: Encoders require their input argument must be uniformly strings or numbers. Got ['float', 'str']

In [143]:
from sklearn.preprocessing import LabelEncoder

df_encoded = df.copy()
le = LabelEncoder()

# Converting all categorical columns to string format to prevent mixed types
for col in cat_features:
    df_encoded[col] = df_encoded[col].astype(str)  # Converting to string before encoding
    df_encoded[col] = le.fit_transform(df_encoded[col])  # Applying Label Encoding

# Save encoded dataset
df_encoded.to_csv("encoded_dataset.csv", index=False)
print("Categorical Features Encoded Successfully.")

Categorical Features Encoded Successfully.


In [145]:
from sklearn.feature_selection import mutual_info_classif

# Computing MI score for categorical features against 'Attack_label'
mi_scores = mutual_info_classif(df_encoded[cat_features], df_encoded["Attack_label"])

# Storing results in a DataFrame
mi_results = pd.DataFrame({'Feature': cat_features, 'MI Score': mi_scores})
mi_results = mi_results.sort_values(by="MI Score", ascending=False)

# Save MI scores
mi_results.to_csv("categorical_feature_importance_mi.csv", index=False)
print("Mutual Information Score Computed and Saved.")

Mutual Information Score Computed and Saved.


In [147]:
from scipy.stats import chi2_contingency

chi2_results = {}

# Computing Chi-Square for each categorical feature
for col in cat_features:
    contingency_table = pd.crosstab(df_encoded[col], df_encoded["Attack_label"])
    chi2, p, _, _ = chi2_contingency(contingency_table)
    chi2_results[col] = p  # Storing p-value

# Converting to DataFrame
chi2_df = pd.DataFrame(list(chi2_results.items()), columns=['Feature', 'P-Value'])
chi2_df = chi2_df.sort_values(by="P-Value")

# Saving Chi-Square results
chi2_df.to_csv("categorical_feature_importance_chi2.csv", index=False)
print("Chi-Square Test Completed. Results Saved.")


Chi-Square Test Completed. Results Saved.


In [151]:
# Selecting features based on MI and Chi-Square significance
significant_features = mi_results[mi_results['MI Score'] > 0.01]['Feature'].tolist()
significant_features += chi2_df[chi2_df['P-Value'] < 0.05]['Feature'].tolist()
significant_features = list(set(significant_features))  # Remove duplicates

print("Selected Features for IDS:", significant_features)

# Saving selected features for final dataset
df_selected = df_encoded[significant_features + ["Attack_label"]]
df_selected.to_csv("final_selected_features.csv", index=False)
print("Final Feature Set Saved.")

Selected Features for IDS: ['http.request.uri.query', 'mqtt.topic', 'mqtt.protoname', 'http.request.version', 'mqtt.conack.flags', 'http.request.method', 'ip.dst_host', 'tcp.srcport', 'tcp.options', 'arp.dst.proto_ipv4', 'ip.src_host', 'arp.src.proto_ipv4']
Final Feature Set Saved.


In [165]:
df.columns

Index(['frame.time', 'ip.src_host', 'ip.dst_host', 'arp.dst.proto_ipv4',
       'arp.opcode', 'arp.hw.size', 'arp.src.proto_ipv4', 'icmp.checksum',
       'icmp.seq_le', 'icmp.transmit_timestamp', 'icmp.unused',
       'http.file_data', 'http.content_length', 'http.request.uri.query',
       'http.request.method', 'http.referer', 'http.request.full_uri',
       'http.request.version', 'http.response', 'http.tls_port', 'tcp.ack',
       'tcp.ack_raw', 'tcp.checksum', 'tcp.connection.fin',
       'tcp.connection.rst', 'tcp.connection.syn', 'tcp.connection.synack',
       'tcp.dstport', 'tcp.flags', 'tcp.flags.ack', 'tcp.len', 'tcp.options',
       'tcp.payload', 'tcp.seq', 'tcp.srcport', 'udp.port', 'udp.stream',
       'udp.time_delta', 'dns.qry.name', 'dns.qry.name.len', 'dns.qry.qu',
       'dns.qry.type', 'dns.retransmission', 'dns.retransmit_request',
       'dns.retransmit_request_in', 'mqtt.conack.flags',
       'mqtt.conflag.cleansess', 'mqtt.conflags', 'mqtt.hdrflags', 'mqtt.len

In [161]:
df_relevant_columns_object_datatype = pd.read_csv("final_selected_features.csv")

In [169]:
df_relevant_columns_object_datatype.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71401 entries, 0 to 71400
Data columns (total 13 columns):
 #   Column                  Non-Null Count  Dtype
---  ------                  --------------  -----
 0   http.request.uri.query  71401 non-null  int64
 1   mqtt.topic              71401 non-null  int64
 2   mqtt.protoname          71401 non-null  int64
 3   http.request.version    71401 non-null  int64
 4   mqtt.conack.flags       71401 non-null  int64
 5   http.request.method     71401 non-null  int64
 6   ip.dst_host             71401 non-null  int64
 7   tcp.srcport             71401 non-null  int64
 8   tcp.options             71401 non-null  int64
 9   arp.dst.proto_ipv4      71401 non-null  int64
 10  ip.src_host             71401 non-null  int64
 11  arp.src.proto_ipv4      71401 non-null  int64
 12  Attack_label            71401 non-null  int64
dtypes: int64(13)
memory usage: 7.1 MB


In [177]:
df_relevant_columns_object_datatype["Attack_label"].unique()

array([0, 1])

In [187]:
df = df.drop(columns=['Attack_type'])

In [189]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71401 entries, 0 to 71400
Data columns (total 50 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   frame.time                 71401 non-null  object 
 1   arp.opcode                 71401 non-null  float64
 2   arp.hw.size                71401 non-null  float64
 3   icmp.checksum              71401 non-null  float64
 4   icmp.seq_le                71401 non-null  float64
 5   icmp.transmit_timestamp    71401 non-null  float64
 6   icmp.unused                71401 non-null  float64
 7   http.file_data             71401 non-null  object 
 8   http.content_length        71401 non-null  float64
 9   http.referer               71401 non-null  object 
 10  http.request.full_uri      71401 non-null  object 
 11  http.response              71401 non-null  float64
 12  http.tls_port              71401 non-null  float64
 13  tcp.ack                    71401 non-null  flo

In [191]:
df = df.drop(columns=['frame.time','http.file_data','http.referer','http.request.full_uri','tcp.payload','mqtt.msg',])

In [193]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71401 entries, 0 to 71400
Data columns (total 44 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   arp.opcode                 71401 non-null  float64
 1   arp.hw.size                71401 non-null  float64
 2   icmp.checksum              71401 non-null  float64
 3   icmp.seq_le                71401 non-null  float64
 4   icmp.transmit_timestamp    71401 non-null  float64
 5   icmp.unused                71401 non-null  float64
 6   http.content_length        71401 non-null  float64
 7   http.response              71401 non-null  float64
 8   http.tls_port              71401 non-null  float64
 9   tcp.ack                    71401 non-null  float64
 10  tcp.ack_raw                71401 non-null  float64
 11  tcp.checksum               71401 non-null  float64
 12  tcp.connection.fin         71401 non-null  float64
 13  tcp.connection.rst         71401 non-null  flo

In [195]:
df_combined = pd.concat([df, df_relevant_columns_object_datatype], axis=1)
df_combined.to_csv("feature_selected.csv", index=False)

In [197]:
df_combined

arp.opcode  arp.hw.size  icmp.checksum  icmp.seq_le  \
0             0.0          0.0            0.0          0.0   
1             0.0          0.0            0.0          0.0   
2             0.0          0.0            0.0          0.0   
3             0.0          0.0            0.0          0.0   
4             0.0          0.0            0.0          0.0   
...           ...          ...            ...          ...   
71396         0.0          0.0        12057.0      57217.0   
71397         0.0          0.0        11289.0      57220.0   
71398         0.0          0.0         7705.0      57234.0   
71399         0.0          0.0         3865.0      57249.0   
71400         0.0          0.0          281.0      57263.0   

       icmp.transmit_timestamp  icmp.unused  http.content_length  \
0                          0.0          0.0                  0.0   
1                          0.0          0.0                  0.0   
2                          0.0          0.0                  0.0   
3                          0.0          0.0                  0.0   
4                          0.0          0.0                  0.0   
...                        ...          ...                  ...   
71396                      0.0          0.0                  0.0   
71397                      0.0          0.0                  0.0   
71398                      0.0          0.0                  0.0   
71399                      0.0          0.0                  0.0   
71400                      0.0          0.0                  0.0   

       http.response  http.tls_port  tcp.ack  ...  http.request.version  \
0                0.0            0.0      1.0  ...                     1   
1                0.0            0.0      1.0  ...                     1   
2                0.0            0.0     15.0  ...                     1   
3                0.0            0.0     15.0  ...                     1   
4                0.0            0.0      5.0  ...                     1   
...              ...            ...      ...  ...                   ...   
71396            0.0            0.0      0.0  ...                     1   
71397            0.0            0.0      0.0  ...                     1   
71398            0.0            0.0      0.0  ...                     1   
71399            0.0            0.0      0.0  ...                     1   
71400            0.0            0.0      0.0  ...                     1   

       mqtt.conack.flags  http.request.method  ip.dst_host  tcp.srcport  \
0                      0                    1         1811         1469   
1                      0                    1         1812        13313   
2                      0                    1         1811         1469   
3                      2                    1         1811         1469   
4                      0                    1         1812        13313   
...                  ...                  ...          ...          ...   
71396                  1                    1         1812            1   
71397                  1                    1         1812            1   
71398                  1                    1         1812            1   
71399                  1                    1         1812            1   
71400                  1                    1         1812            1   

       tcp.options  arp.dst.proto_ipv4  ip.src_host  arp.src.proto_ipv4  \
0            28231                   0         2451                   0   
1                0                   0         2450                   0   
2                0                   0         2451                   0   
3                0                   0         2451                   0   
4                0                   0         2450                   0   
...            ...                 ...          ...                 ...   
71396            1                   0         1493                   1   
71397            1                   0     

In [199]:
df_combined.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71401 entries, 0 to 71400
Data columns (total 57 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   arp.opcode                 71401 non-null  float64
 1   arp.hw.size                71401 non-null  float64
 2   icmp.checksum              71401 non-null  float64
 3   icmp.seq_le                71401 non-null  float64
 4   icmp.transmit_timestamp    71401 non-null  float64
 5   icmp.unused                71401 non-null  float64
 6   http.content_length        71401 non-null  float64
 7   http.response              71401 non-null  float64
 8   http.tls_port              71401 non-null  float64
 9   tcp.ack                    71401 non-null  float64
 10  tcp.ack_raw                71401 non-null  float64
 11  tcp.checksum               71401 non-null  float64
 12  tcp.connection.fin         71401 non-null  float64
 13  tcp.connection.rst         71401 non-null  flo

In [201]:
# Identifing duplicate columns
duplicate_columns = df.columns[df.columns.duplicated()].tolist()
print("Duplicate Columns:", duplicate_columns)

# Droping duplicate columns
df = df.loc[:, ~df.columns.duplicated()]

# Saving the cleaned dataset
df.to_csv("feature_selected_cleaned.csv", index=False)
print("Duplicate columns removed. Cleaned dataset saved.")

Duplicate Columns: []
Duplicate columns removed. Cleaned dataset saved.


In [203]:
df_fs = pd.read_csv("feature_selected_cleaned.csv")

In [205]:
df_fs

arp.opcode  arp.hw.size  icmp.checksum  icmp.seq_le  \
0             0.0          0.0            0.0          0.0   
1             0.0          0.0            0.0          0.0   
2             0.0          0.0            0.0          0.0   
3             0.0          0.0            0.0          0.0   
4             0.0          0.0            0.0          0.0   
...           ...          ...            ...          ...   
71396         0.0          0.0        12057.0      57217.0   
71397         0.0          0.0        11289.0      57220.0   
71398         0.0          0.0         7705.0      57234.0   
71399         0.0          0.0         3865.0      57249.0   
71400         0.0          0.0          281.0      57263.0   

       icmp.transmit_timestamp  icmp.unused  http.content_length  \
0                          0.0          0.0                  0.0   
1                          0.0          0.0                  0.0   
2                          0.0          0.0                  0.0   
3                          0.0          0.0                  0.0   
4                          0.0          0.0                  0.0   
...                        ...          ...                  ...   
71396                      0.0          0.0                  0.0   
71397                      0.0          0.0                  0.0   
71398                      0.0          0.0                  0.0   
71399                      0.0          0.0                  0.0   
71400                      0.0          0.0                  0.0   

       http.response  http.tls_port  tcp.ack  ...  mqtt.len  \
0                0.0            0.0      1.0  ...       0.0   
1                0.0            0.0      1.0  ...      12.0   
2                0.0            0.0     15.0  ...       0.0   
3                0.0            0.0     15.0  ...       2.0   
4                0.0            0.0      5.0  ...      39.0   
...              ...            ...      ...  ...       ...   
71396            0.0            0.0      0.0  ...       0.0   
71397            0.0            0.0      0.0  ...       0.0   
71398            0.0            0.0      0.0  ...       0.0   
71399            0.0            0.0      0.0  ...       0.0   
71400            0.0            0.0      0.0  ...       0.0   

       mqtt.msg_decoded_as  mqtt.msgtype  mqtt.proto_len  mqtt.topic_len  \
0                      0.0           0.0             0.0             0.0   
1                      0.0           1.0             4.0             0.0   
2                      0.0           0.0             0.0             0.0   
3                      0.0           2.0             0.0             0.0   
4                      0.0           3.0             0.0            24.0   
...                    ...           ...             ...             ...   
71396                  0.0           0.0             0.0             0.0   
71397                  0.0           0.0             0.0             0.0   
71398                  0.0           0.0             0.0             0.0   
71399                  0.0           0.0             0.0             0.0   
71400                  0.0           0.0             0.0             0.0   

       mqtt.ver  mbtcp.len  mbtcp.trans_id  mbtcp.unit_id  Attack_label  
0           0.0        0.0             0.0            0.0             0  
1           4.0        0.0             0.0            0.0             0  
2           0.0        0.0             0.0            0.0             0  
3           0.0        0.0             0.0            0.0             0  
4           0.0        0.0             0.0            0.0             0  
...         ...        ...             ...            ...           ...  
71396       0.0        0.0             0.0            0.0             1  
71397       0.0        0.0             0.0            0.0             1  
71398       0.0        0.0             0.0            0.0             1  
71399       0.0        0.0         

In [215]:
df_fs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71401 entries, 0 to 71400
Data columns (total 44 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   arp.opcode                 71401 non-null  float64
 1   arp.hw.size                71401 non-null  float64
 2   icmp.checksum              71401 non-null  float64
 3   icmp.seq_le                71401 non-null  float64
 4   icmp.transmit_timestamp    71401 non-null  float64
 5   icmp.unused                71401 non-null  float64
 6   http.content_length        71401 non-null  float64
 7   http.response              71401 non-null  float64
 8   http.tls_port              71401 non-null  float64
 9   tcp.ack                    71401 non-null  float64
 10  tcp.ack_raw                71401 non-null  float64
 11  tcp.checksum               71401 non-null  float64
 12  tcp.connection.fin         71401 non-null  float64
 13  tcp.connection.rst         71401 non-null  flo

### **Feature Engineering Progress Report**  

#### **1. Dataset Cleaning & Feature Selection**  
- Processed the **EdgeIoT2023** dataset and performed feature selection.  
- Removed **redundant and irrelevant categorical features** based on **Mutual Information (MI) Scores** and **Chi-Square Test**.  
- Eliminated **duplicate columns** to avoid redundancy and reduce memory usage.  

---

#### **2. Current Dataset Overview**  
- **Total Entries:** 71,401  
- **Total Features After Cleaning:** 44  
- **Feature Types:**  
  - **43 Float Columns** (Numeric Network Traffic Features)  
  - **1 Integer Column** (`Attack_label`, Target Variable)  

---

#### **3. Improvements Achieved**  
 **Reduced Memory Usage** → Optimised dataset from **34.3MB to 24.0MB**.  
 **Removed Unnecessary Features** → Ensures **faster model training** and **better generalisation**.  
 **Optimised Feature Set for IDS** → Dataset now contains only **high-impact network features** for intrusion detection.  

---

#### **4. Next Steps**  
- Perform **feature scaling** and **normalisation** to prepare the dataset for model training.  
- Train the **Layer 1 Autoencoder-LSTM** for **anomaly detection**.  
- Evaluate **reconstruction error-based anomaly detection** before forwarding data to **Layer 2** for classification.  

With these refinements, the dataset is now **optimised for real-time intrusion detection in IoT environments**.  


In [219]:
df_fs.describe()

arp.opcode   arp.hw.size  icmp.checksum   icmp.seq_le  \
count  71401.000000  71401.000000   71401.000000  71401.000000   
mean       0.020308      0.093108    2573.315612   2275.936429   
std        0.171684      0.741611   10318.127260   8980.693684   
min        0.000000      0.000000       0.000000      0.000000   
25%        0.000000      0.000000       0.000000      0.000000   
50%        0.000000      0.000000       0.000000      0.000000   
75%        0.000000      0.000000       0.000000      0.000000   
max        2.000000      6.000000   65513.000000  65036.000000   

       icmp.transmit_timestamp  icmp.unused  http.content_length  \
count             7.140100e+04      71401.0         71401.000000   
mean              8.943678e+04          0.0            15.005322   
std               2.621692e+06          0.0            96.106592   
min               0.000000e+00          0.0             0.000000   
25%               0.000000e+00          0.0             0.000000   
50%               0.000000e+00          0.0             0.000000   
75%               0.000000e+00          0.0             0.000000   
max               7.728902e+07          0.0          1824.000000   

       http.response  http.tls_port       tcp.ack  ...      mqtt.len  \
count   71401.000000        71401.0  7.140100e+04  ...  71401.000000   
mean        0.047660            0.0  7.482104e+07  ...      0.459475   
std         0.213048            0.0  3.194048e+08  ...      3.776006   
min         0.000000            0.0  0.000000e+00  ...      0.000000   
25%         0.000000            0.0  0.000000e+00  ...      0.000000   
50%         0.000000            0.0  5.000000e+00  ...      0.000000   
75%         0.000000            0.0  4.410000e+02  ...      0.000000   
max         1.000000            0.0  2.146757e+09  ...     39.000000   

       mqtt.msg_decoded_as  mqtt.msgtype  mqtt.proto_len  mqtt.topic_len  \
count              71401.0  71401.000000    71401.000000    71401.000000   
mean                   0.0      0.173387        0.034677        0.208064   
std                    0.0      1.338105        0.370822        2.224932   
min                    0.0      0.000000        0.000000        0.000000   
25%                    0.0      0.000000        0.000000        0.000000   
50%                    0.0      0.000000        0.000000        0.000000   
75%                    0.0      0.000000        0.000000        0.000000   
max                    0.0     14.000000        4.000000       24.000000   

           mqtt.ver  mbtcp.len  mbtcp.trans_id  mbtcp.unit_id  Attack_label  
count  71401.000000    71401.0         71401.0        71401.0  71401.000000  
mean       0.034677        0.0             0.0            0.0      0.859946  
std        0.370822        0.0             0.0            0.0      0.347046  
min        0.000000        0.0             0.0            0.0      0.000000  
25%        0.000000        0.0             0.0            0.0      1.000000  
50%        0.000000        0.0             0.0            0.0      1.000000  
75%        0.000000        0.0             0.0            0.0      1.000000  
max        4.000000        0.0             0.0            0.0      1.000000  

[8 rows x 44 columns]

In [223]:
# Getting summary statistics of the dataset
df_fs.describe().to_csv("data_distribution_statistics.csv")

# Displaying summary statistics
df_fs.describe()

arp.opcode   arp.hw.size  icmp.checksum   icmp.seq_le  \
count  71401.000000  71401.000000   71401.000000  71401.000000   
mean       0.020308      0.093108    2573.315612   2275.936429   
std        0.171684      0.741611   10318.127260   8980.693684   
min        0.000000      0.000000       0.000000      0.000000   
25%        0.000000      0.000000       0.000000      0.000000   
50%        0.000000      0.000000       0.000000      0.000000   
75%        0.000000      0.000000       0.000000      0.000000   
max        2.000000      6.000000   65513.000000  65036.000000   

       icmp.transmit_timestamp  icmp.unused  http.content_length  \
count             7.140100e+04      71401.0         71401.000000   
mean              8.943678e+04          0.0            15.005322   
std               2.621692e+06          0.0            96.106592   
min               0.000000e+00          0.0             0.000000   
25%               0.000000e+00          0.0             0.000000   
50%               0.000000e+00          0.0             0.000000   
75%               0.000000e+00          0.0             0.000000   
max               7.728902e+07          0.0          1824.000000   

       http.response  http.tls_port       tcp.ack  ...      mqtt.len  \
count   71401.000000        71401.0  7.140100e+04  ...  71401.000000   
mean        0.047660            0.0  7.482104e+07  ...      0.459475   
std         0.213048            0.0  3.194048e+08  ...      3.776006   
min         0.000000            0.0  0.000000e+00  ...      0.000000   
25%         0.000000            0.0  0.000000e+00  ...      0.000000   
50%         0.000000            0.0  5.000000e+00  ...      0.000000   
75%         0.000000            0.0  4.410000e+02  ...      0.000000   
max         1.000000            0.0  2.146757e+09  ...     39.000000   

       mqtt.msg_decoded_as  mqtt.msgtype  mqtt.proto_len  mqtt.topic_len  \
count              71401.0  71401.000000    71401.000000    71401.000000   
mean                   0.0      0.173387        0.034677        0.208064   
std                    0.0      1.338105        0.370822        2.224932   
min                    0.0      0.000000        0.000000        0.000000   
25%                    0.0      0.000000        0.000000        0.000000   
50%                    0.0      0.000000        0.000000        0.000000   
75%                    0.0      0.000000        0.000000        0.000000   
max                    0.0     14.000000        4.000000       24.000000   

           mqtt.ver  mbtcp.len  mbtcp.trans_id  mbtcp.unit_id  Attack_label  
count  71401.000000    71401.0         71401.0        71401.0  71401.000000  
mean       0.034677        0.0             0.0            0.0      0.859946  
std        0.370822        0.0             0.0            0.0      0.347046  
min        0.000000        0.0             0.0            0.0      0.000000  
25%        0.000000        0.0             0.0            0.0      1.000000  
50%        0.000000        0.0             0.0            0.0      1.000000  
75%        0.000000        0.0             0.0            0.0      1.000000  
max        4.000000        0.0             0.0            0.0      1.000000  

[8 rows x 44 columns]

In [227]:
from sklearn.preprocessing import MinMaxScaler
import pickle

# Initializing Min-Max Scaler
scaler = MinMaxScaler()

# Fitting and transforming the dataset
df_fs_scaled = pd.DataFrame(scaler.fit_transform(df_fs), columns=df_fs.columns)

# Saving the scaler for future use during inference
with open("minmax_scaler.pkl", "wb") as f:
    pickle.dump(scaler, f)

# Saving the scaled dataset
df_fs_scaled.to_csv("df_fs_scaled.csv", index=False)

print("Done")

Done


In [229]:
df_fs_scaled.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71401 entries, 0 to 71400
Data columns (total 44 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   arp.opcode                 71401 non-null  float64
 1   arp.hw.size                71401 non-null  float64
 2   icmp.checksum              71401 non-null  float64
 3   icmp.seq_le                71401 non-null  float64
 4   icmp.transmit_timestamp    71401 non-null  float64
 5   icmp.unused                71401 non-null  float64
 6   http.content_length        71401 non-null  float64
 7   http.response              71401 non-null  float64
 8   http.tls_port              71401 non-null  float64
 9   tcp.ack                    71401 non-null  float64
 10  tcp.ack_raw                71401 non-null  float64
 11  tcp.checksum               71401 non-null  float64
 12  tcp.connection.fin         71401 non-null  float64
 13  tcp.connection.rst         71401 non-null  flo

In [231]:
df_fs_scaled.describe().to_csv("data_distribution_statistics_after_fs.csv")

In [233]:
df_fs_scaled.describe()

arp.opcode   arp.hw.size  icmp.checksum   icmp.seq_le  \
count  71401.000000  71401.000000   71401.000000  71401.000000   
mean       0.010154      0.015518       0.039279      0.034995   
std        0.085842      0.123602       0.157497      0.138088   
min        0.000000      0.000000       0.000000      0.000000   
25%        0.000000      0.000000       0.000000      0.000000   
50%        0.000000      0.000000       0.000000      0.000000   
75%        0.000000      0.000000       0.000000      0.000000   
max        1.000000      1.000000       1.000000      1.000000   

       icmp.transmit_timestamp  icmp.unused  http.content_length  \
count             71401.000000      71401.0         71401.000000   
mean                  0.001157          0.0             0.008227   
std                   0.033921          0.0             0.052690   
min                   0.000000          0.0             0.000000   
25%                   0.000000          0.0             0.000000   
50%                   0.000000          0.0             0.000000   
75%                   0.000000          0.0             0.000000   
max                   1.000000          0.0             1.000000   

       http.response  http.tls_port       tcp.ack  ...      mqtt.len  \
count   71401.000000        71401.0  7.140100e+04  ...  71401.000000   
mean        0.047660            0.0  3.485306e-02  ...      0.011781   
std         0.213048            0.0  1.487848e-01  ...      0.096821   
min         0.000000            0.0  0.000000e+00  ...      0.000000   
25%         0.000000            0.0  0.000000e+00  ...      0.000000   
50%         0.000000            0.0  2.329095e-09  ...      0.000000   
75%         0.000000            0.0  2.054262e-07  ...      0.000000   
max         1.000000            0.0  1.000000e+00  ...      1.000000   

       mqtt.msg_decoded_as  mqtt.msgtype  mqtt.proto_len  mqtt.topic_len  \
count              71401.0  71401.000000    71401.000000    71401.000000   
mean                   0.0      0.012385        0.008669        0.008669   
std                    0.0      0.095579        0.092705        0.092705   
min                    0.0      0.000000        0.000000        0.000000   
25%                    0.0      0.000000        0.000000        0.000000   
50%                    0.0      0.000000        0.000000        0.000000   
75%                    0.0      0.000000        0.000000        0.000000   
max                    0.0      1.000000        1.000000        1.000000   

           mqtt.ver  mbtcp.len  mbtcp.trans_id  mbtcp.unit_id  Attack_label  
count  71401.000000    71401.0         71401.0        71401.0  71401.000000  
mean       0.008669        0.0             0.0            0.0      0.859946  
std        0.092705        0.0             0.0            0.0      0.347046  
min        0.000000        0.0             0.0            0.0      0.000000  
25%        0.000000        0.0             0.0            0.0      1.000000  
50%        0.000000        0.0             0.0            0.0      1.000000  
75%        0.000000        0.0             0.0            0.0      1.000000  
max        1.000000        0.0             0.0            0.0      1.000000  

[8 rows x 44 columns]

In [235]:
df_fs_scaled.head()

arp.opcode  arp.hw.size  icmp.checksum  icmp.seq_le  \
0         0.0          0.0            0.0          0.0   
1         0.0          0.0            0.0          0.0   
2         0.0          0.0            0.0          0.0   
3         0.0          0.0            0.0          0.0   
4         0.0          0.0            0.0          0.0   

   icmp.transmit_timestamp  icmp.unused  http.content_length  http.response  \
0                      0.0          0.0                  0.0            0.0   
1                      0.0          0.0                  0.0            0.0   
2                      0.0          0.0                  0.0            0.0   
3                      0.0          0.0                  0.0            0.0   
4                      0.0          0.0                  0.0            0.0   

   http.tls_port       tcp.ack  ...  mqtt.len  mqtt.msg_decoded_as  \
0            0.0  4.658190e-10  ...  0.000000                  0.0   
1            0.0  4.658190e-10  ...  0.307692                  0.0   
2            0.0  6.987285e-09  ...  0.000000                  0.0   
3            0.0  6.987285e-09  ...  0.051282                  0.0   
4            0.0  2.329095e-09  ...  1.000000                  0.0   

   mqtt.msgtype  mqtt.proto_len  mqtt.topic_len  mqtt.ver  mbtcp.len  \
0      0.000000             0.0             0.0       0.0        0.0   
1      0.071429             1.0             0.0       1.0        0.0   
2      0.000000             0.0             0.0       0.0        0.0   
3      0.142857             0.0             0.0       0.0        0.0   
4      0.214286             0.0             1.0       0.0        0.0   

   mbtcp.trans_id  mbtcp.unit_id  Attack_label  
0             0.0            0.0           0.0  
1             0.0            0.0           0.0  
2             0.0            0.0           0.0  
3             0.0            0.0           0.0  
4             0.0            0.0           0.0  

[5 rows x 44 columns]

In [237]:
df_fs_scaled.to_csv("training_dataset.csv")